Adopt more spans in ImageBackingStore.h

cdumez opened the pull request:
#### 1db6fc53597cc71b6be16c96fe3ba78a08a6a980
<pre>
Adopt more spans in ImageBackingStore.h
<a href="https://bugs.webkit.org/show_bug.cgi?id=285389">https://bugs.webkit.org/show_bug.cgi?id=285389</a>

Reviewed by Darin Adler.

* Source/WebCore/platform/graphics/ImageBackingStore.h:
(WebCore::ImageBackingStore::setSize):
(WebCore::ImageBackingStore::clear):
(WebCore::ImageBackingStore::clearRect):
(WebCore::ImageBackingStore::fillRect):
(WebCore::ImageBackingStore::repeatFirstRow):
(WebCore::ImageBackingStore::pixelAt const):
(WebCore::ImageBackingStore::setPixel):
(WebCore::ImageBackingStore::blendPixel):
(WebCore::ImageBackingStore::ImageBackingStore):
* Source/WebCore/platform/image-decoders/avif/AVIFImageReader.cpp:
(WebCore::AVIFImageReader::decodeFrame):
* Source/WebCore/platform/image-decoders/gif/GIFImageDecoder.cpp:
(WebCore::GIFImageDecoder::haveDecodedRow):
* Source/WebCore/platform/image-decoders/jpeg/JPEGImageDecoder.cpp:
(WebCore::setPixel):
(WebCore::JPEGImageDecoder::outputScanlines):
* Source/WebCore/platform/image-decoders/jpegxl/JPEGXLImageDecoder.cpp:
(WebCore::JPEGXLImageDecoder::imageOut):
* Source/WebCore/platform/image-decoders/png/PNGImageDecoder.cpp:
(WebCore::PNGImageDecoder::rowAvailable):
(WebCore::PNGImageDecoder::frameComplete):
* Source/WebCore/platform/image-decoders/webp/WEBPImageDecoder.cpp:
(WebCore::WEBPImageDecoder::applyPostProcessing):

Canonical link: <a href="https://commits.webkit.org/288455@main">https://commits.webkit.org/288455@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/5f65c5a9bebd26a4e4a63e596b5b8553a453061f

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/83447 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/3064 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/37736 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/88523 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/34457 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/85535 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/3149 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/11021 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/64918 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/22658 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/86497 "Passed tests") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/132/builds/2302 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/75829 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/45205 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/2204 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/30035 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/33507 "Built successfully") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/13/builds/73302 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/30768 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/89899 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/10712 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/7723 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/73343 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/10937 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/71656 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/72571 "Passed tests") | 
| | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/16802 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/119/builds/15527 "Passed tests") | [  ~~🛠 playstation~~](https://ews-build.webkit.org/#/builders/134/builds/2045 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/12878 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/10667 "Built successfully") | [  ~~🛠 mac-safer-cpp~~](https://ews-build.webkit.org/#/builders/120/builds/16136 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/10518 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/13989 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/12289 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

darinadler commented on Source/WebCore/platform/graphics/ImageBackingStore.h:
```diff
@@ -89,19 +88,18 @@ class ImageBackingStore {
-            memset(start, 0, rowBytes);
-            start += m_size.width();
+            zeroSpan(start.first(rect.width()));
+            start = start.subspan(m_size.width());
```
```suggestion
            skip(start, m_size.width());
```

darinadler commented on Source/WebCore/platform/graphics/ImageBackingStore.h:
```diff
@@ -110,12 +108,12 @@ class ImageBackingStore {
             for (int j = 0; j < rect.width(); ++j)
                 start[j] = pixelValue;
-            start += m_size.width();
+            start = start.subspan(m_size.width());
```
```suggestion
            skip(start, m_size.width());
```

darinadler commented on Source/WebCore/platform/graphics/ImageBackingStore.h:
```diff
@@ -124,38 +122,36 @@ class ImageBackingStore {
-            memcpy(dest, src, rowBytes);
-            dest += m_size.width();
+            memcpySpan(dest, src.first(rect.width()));
+            dest = dest.subspan(m_size.width());
```
```suggestion
            skip(dest, m_size.width());
```

darinadler commented on Source/WebCore/platform/graphics/ImageBackingStore.h:
```diff
@@ -124,38 +122,36 @@ class ImageBackingStore {
     }
 
-    uint32_t* pixelAt(int x, int y) const
+    std::span<uint32_t> pixelAt(int x, int y) const
```
This function name made more sense when it returned only a pointer. But I don't suppose I have a better name. Maybe with the word _after_ in it?

cdumez replied:
Proposing `pixelsStartingAt()` which returns a `std::span<uint32_t>` and `pixelAt()` which returns a `uint32_t&`

darinadler commented on Source/WebCore/platform/image-decoders/gif/GIFImageDecoder.cpp:
```diff
@@ -253,9 +251,9 @@ bool GIFImageDecoder::haveDecodedRow(unsigned frameIndex, const Vector<unsigned
+                buffer.backingStore()->setPixel(currentAddress[0], 0, 0, 0, 0);
         }
-        ++currentAddress;
+        currentAddress = currentAddress.subspan(1);
```
```suggestion
        skip(currentAddress, 1);
```

darinadler commented on Source/WebCore/platform/image-decoders/jpeg/JPEGImageDecoder.cpp:
```diff
@@ -629,18 +629,18 @@ bool JPEGImageDecoder::outputScanlines(ScalableImageDecoderFrame& buffer)
         for (int x = 0; x < width; ++x) {
             setPixel<colorSpace>(buffer, currentAddress, samples, x);
-            ++currentAddress;
+            currentAddress = currentAddress.subspan(1);
```
```suggestion
            skip(currentAddress, 1);
```

darinadler commented on Source/WebCore/platform/image-decoders/jpeg/JPEGImageDecoder.cpp:
```diff
@@ -667,7 +667,7 @@ bool JPEGImageDecoder::outputScanlines()
     if (turboSwizzled(info->out_color_space)) {
         while (info->output_scanline < info->output_height) {
-            unsigned char* row = reinterpret_cast<unsigned char*>(buffer.backingStore()->pixelAt(0, info->output_scanline));
+            unsigned char* row = reinterpret_cast<unsigned char*>(buffer.backingStore()->pixelAt(0, info->output_scanline).data());
```
```suggestion
            auto* row = reinterpret_cast<unsigned char*>(buffer.backingStore()->pixelAt(0, info->output_scanline).data());
```

cdumez committed:
Adopt more spans in ImageBackingStore.h
https://bugs.webkit.org/show_bug.cgi?id=285389

Reviewed by Darin Adler.

* Source/WebCore/platform/graphics/ImageBackingStore.h:
(WebCore::ImageBackingStore::setSize):
(WebCore::ImageBackingStore::clear):
(WebCore::ImageBackingStore::clearRect):
(WebCore::ImageBackingStore::fillRect):
(WebCore::ImageBackingStore::repeatFirstRow):
(WebCore::ImageBackingStore::pixelAt const):
(WebCore::ImageBackingStore::setPixel):
(WebCore::ImageBackingStore::blendPixel):
(WebCore::ImageBackingStore::ImageBackingStore):
* Source/WebCore/platform/image-decoders/avif/AVIFImageReader.cpp:
(WebCore::AVIFImageReader::decodeFrame):
* Source/WebCore/platform/image-decoders/gif/GIFImageDecoder.cpp:
(WebCore::GIFImageDecoder::haveDecodedRow):
* Source/WebCore/platform/image-decoders/jpeg/JPEGImageDecoder.cpp:
(WebCore::setPixel):
(WebCore::JPEGImageDecoder::outputScanlines):
* Source/WebCore/platform/image-decoders/jpegxl/JPEGXLImageDecoder.cpp:
(WebCore::JPEGXLImageDecoder::imageOut):
* Source/WebCore/platform/image-decoders/png/PNGImageDecoder.cpp:
(WebCore::PNGImageDecoder::rowAvailable):
(WebCore::PNGImageDecoder::frameComplete):
* Source/WebCore/platform/image-decoders/webp/WEBPImageDecoder.cpp:
(WebCore::WEBPImageDecoder::applyPostProcessing):

Canonical link: https://commits.webkit.org/288455@main
